File: MiniDB_Projects/Team_candyman/internal/storage/heapfile.go
package storage

import "encoding/binary"

// RID (record id) locates a record: a page and a slot within it. RIDs are stable
// across the record's lifetime because deletes tombstone slots rather than moving
// records. The B+Tree primary-key index maps key -> RID.
type RID struct {
	PageID PageID
	Slot   uint16
}

// Encode/DecodeRID serialize a RID to 6 bytes for storage as a B+Tree value.
func (r RID) Encode() []byte {
	b := make([]byte, 6)
	binary.BigEndian.PutUint32(b[0:], uint32(r.PageID))
	binary.BigEndian.PutUint16(b[4:], r.Slot)
	return b
}

func DecodeRID(b []byte) RID {
	return RID{
		PageID: PageID(binary.BigEndian.Uint32(b[0:])),
		Slot:   binary.BigEndian.Uint16(b[4:]),
	}
}

// HeapFile is an unordered collection of records spread across a chain of pages
// in the data file. Each table owns one HeapFile rooted at firstPage.
type HeapFile struct {
	bp        *BufferPool
	firstPage PageID
}

// NewHeapFile allocates the first page of a new heap and returns the heap plus
// its root page id (the catalog persists this id).
func NewHeapFile(bp *BufferPool) (*HeapFile, PageID, error) {
	id, _, err := bp.NewPage(FileData)
	if err != nil {
		return nil, 0, err
	}
	bp.Unpin(FileData, id, true)
	return &HeapFile{bp: bp, firstPage: id}, id, nil
}

// OpenHeapFile reopens an existing heap rooted at firstPage.
func OpenHeapFile(bp *BufferPool, firstPage PageID) *HeapFile {
	return &HeapFile{bp: bp, firstPage: firstPage}
}

// Insert appends a record, allocating and linking a new page if the chain is full.
// It returns the new record's RID and the LSN slot is left to the caller (the WAL
// layer sets pageLSN via SetPageLSN after logging).
func (h *HeapFile) Insert(rec []byte) (RID, error) {
	cur := h.firstPage
	for {
		page, err := h.bp.Fetch(FileData, cur)
		if err != nil {
			return RID{}, err
		}
		slot, err := page.Insert(rec)
		if err == nil {
			h.bp.Unpin(FileData, cur, true)
			return RID{PageID: cur, Slot: uint16(slot)}, nil
		}
		next := page.NextPage()
		if next != InvalidPageID {
			h.bp.Unpin(FileData, cur, false)
			cur = next
			continue
		}
		// end of chain and record didn't fit: append a fresh page
		newID, newPage, err := h.bp.NewPage(FileData)
		if err != nil {
			h.bp.Unpin(FileData, cur, false)
			return RID{}, err
		}
		slot, err = newPage.Insert(rec)
		if err != nil {
			h.bp.Unpin(FileData, newID, false)
			h.bp.Unpin(FileData, cur, false)
			return RID{}, err
		}
		page.SetNextPage(newID)
		h.bp.Unpin(FileData, cur, true)
		h.bp.Unpin(FileData, newID, true)
		return RID{PageID: newID, Slot: uint16(slot)}, nil
	}
}

// InsertAt writes a record at a specific RID's page+slot during redo recovery.
// It only applies the write if the page's LSN is older than recLSN (idempotent
// redo). Used by the WAL recovery path.
func (h *HeapFile) InsertAt(rid RID, rec []byte, recLSN uint64) error {
	page, err := h.bp.Fetch(FileData, rid.PageID)
	if err != nil {
		return err
	}
	defer func() { h.bp.Unpin(FileData, rid.PageID, true) }()
	if page.LSN() >= recLSN {
		return nil // already reflects this update
	}
	// best-effort: append; slot may differ but redo replays full history in order
	if _, err := page.Insert(rec); err != nil {
		return err
	}
	page.SetLSN(recLSN)
	return nil
}

// Read returns the record bytes at a RID.
func (h *HeapFile) Read(rid RID) ([]byte, error) {
	page, err := h.bp.Fetch(FileData, rid.PageID)
	if err != nil {
		return nil, err
	}
	defer h.bp.Unpin(FileData, rid.PageID, false)
	return page.Read(int(rid.Slot))
}

// Delete tombstones the record at a RID and bumps its page LSN.
func (h *HeapFile) Delete(rid RID, lsn uint64) error {
	page, err := h.bp.Fetch(FileData, rid.PageID)
	if err != nil {
		return err
	}
	defer func() { h.bp.Unpin(FileData, rid.PageID, true) }()
	if err := page.Delete(int(rid.Slot)); err != nil {
		return err
	}
	if lsn > page.LSN() {
		page.SetLSN(lsn)
	}
	return nil
}

// SetPageLSN records the LSN of the log record that produced the latest change
// to the page holding rid (used by the WAL layer right after Insert).
func (h *HeapFile) SetPageLSN(pid PageID, lsn uint64) error {
	page, err := h.bp.Fetch(FileData, pid)
	if err != nil {
		return err
	}
	defer h.bp.Unpin(FileData, pid, true)
	if lsn > page.LSN() {
		page.SetLSN(lsn)
	}
	return nil
}

// ScanFunc is called for every live record during a scan. Returning false stops.
type ScanFunc func(rid RID, rec []byte) bool

// Scan visits every live (non-tombstone) record in the heap.
func (h *HeapFile) Scan(fn ScanFunc) error {
	cur := h.firstPage
	for cur != InvalidPageID {
		page, err := h.bp.Fetch(FileData, cur)
		if err != nil {
			return err
		}
		n := page.SlotCount()
		next := page.NextPage()
		for s := 0; s < n; s++ {
			if page.IsTombstone(s) {
				continue
			}
			rec, err := page.Read(s)
			if err != nil {
				continue
			}
			if !fn(RID{PageID: cur, Slot: uint16(s)}, rec) {
				h.bp.Unpin(FileData, cur, false)
				return nil
			}
		}
		h.bp.Unpin(FileData, cur, false)
		cur = next
	}
	return nil
}

// FirstPage returns the heap's root page id (for catalog persistence).
func (h *HeapFile) FirstPage() PageID { return h.firstPage }

File: MiniDB_Projects/Team_Ayush/src/storage/heap_file.cpp
#include "storage/heap_file.h"

#include <cstring>

namespace minidb {

namespace {
const int kHeaderSize = 8;

PageId GetNext(const char* p) {
  int32_t v; std::memcpy(&v, p, 4); return v;
}
void SetNext(char* p, PageId v) { std::memcpy(p, &v, 4); }

void SetRecSize(char* p, int16_t v) { std::memcpy(p + 4, &v, 2); }

int16_t GetNumUsed(const char* p) {
  int16_t v; std::memcpy(&v, p + 6, 2); return v;
}
void SetNumUsed(char* p, int16_t v) { std::memcpy(p + 6, &v, 2); }

void InitPage(char* p, int record_size) {
  std::memset(p, 0, PAGE_SIZE);
  SetNext(p, INVALID_PAGE_ID);
  SetRecSize(p, static_cast<int16_t>(record_size));
  SetNumUsed(p, 0);
}
}  // namespace

int HeapFile::SlotsPerPage() const {
  return (PAGE_SIZE - kHeaderSize) / (1 + record_size_);
}

PageId HeapFile::CreateNew(BufferPool* bp, int record_size) {
  PageId pid;
  Frame* f = bp->NewPage(&pid);
  InitPage(f->data, record_size);
  bp->UnpinPage(pid, true);
  return pid;
}

RID HeapFile::Insert(const char* record) {
  const int slot_stride = 1 + record_size_;
  const int slots = SlotsPerPage();

  PageId pid = first_page_;
  PageId last = first_page_;
  while (pid != INVALID_PAGE_ID) {
    Frame* f = bp_->FetchPage(pid);
    char* p = f->data;
    for (int i = 0; i < slots; ++i) {
      char* slot = p + kHeaderSize + i * slot_stride;
      if (slot[0] == 0) {
        slot[0] = 1;
        std::memcpy(slot + 1, record, record_size_);
        SetNumUsed(p, static_cast<int16_t>(GetNumUsed(p) + 1));
        bp_->UnpinPage(pid, true);
        return RID(pid, static_cast<int16_t>(i));
      }
    }
    last = pid;
    PageId next = GetNext(p);
    bp_->UnpinPage(pid, false);
    pid = next;
  }

  // No free slot anywhere: append a new page and link it in.
  PageId np;
  Frame* nf = bp_->NewPage(&np);
  char* p = nf->data;
  InitPage(p, record_size_);
  char* slot = p + kHeaderSize;  // slot 0
  slot[0] = 1;
  std::memcpy(slot + 1, record, record_size_);
  SetNumUsed(p, 1);
  bp_->UnpinPage(np, true);

  Frame* lf = bp_->FetchPage(last);
  SetNext(lf->data, np);
  bp_->UnpinPage(last, true);

  return RID(np, 0);
}

bool HeapFile::Get(RID rid, char* out) {
  Frame* f = bp_->FetchPage(rid.page);
  char* slot = f->data + kHeaderSize + rid.slot * (1 + record_size_);
  bool live = slot[0] == 1;
  if (live) std::memcpy(out, slot + 1, record_size_);
  bp_->UnpinPage(rid.page, false);
  return live;
}

bool HeapFile::Update(RID rid, const char* record) {
  Frame* f = bp_->FetchPage(rid.page);
  char* slot = f->data + kHeaderSize + rid.slot * (1 + record_size_);
  bool live = slot[0] == 1;
  if (live) std::memcpy(slot + 1, record, record_size_);
  bp_->UnpinPage(rid.page, live);
  return live;
}

bool HeapFile::Delete(RID rid) {
  Frame* f = bp_->FetchPage(rid.page);
  char* p = f->data;
  char* slot = p + kHeaderSize + rid.slot * (1 + record_size_);
  bool live = slot[0] == 1;
  if (live) {
    slot[0] = 0;
    SetNumUsed(p, static_cast<int16_t>(GetNumUsed(p) - 1));
  }
  bp_->UnpinPage(rid.page, live);
  return live;
}

void HeapFile::Scan(const std::function<void(RID, const char*)>& visitor) {
  const int slot_stride = 1 + record_size_;
  const int slots = SlotsPerPage();
  PageId pid = first_page_;
  while (pid != INVALID_PAGE_ID) {
    Frame* f = bp_->FetchPage(pid);
    char* p = f->data;
    for (int i = 0; i < slots; ++i) {
      char* slot = p + kHeaderSize + i * slot_stride;
      if (slot[0] == 1) {
        visitor(RID(pid, static_cast<int16_t>(i)), slot + 1);
      }
    }
    PageId next = GetNext(p);
    bp_->UnpinPage(pid, false);
    pid = next;
  }
}

}  // namespace minidb
